toml-based example (#238)
CliMA/EnsembleKalmanProcesses.jl: docs/src/examples/sinusoid_example_toml.md

# TOML interface for fitting parameters of a sinusoid

Here we revisit the [simple example](@ref sinusoid-example). 

The purpose of this example is to demonstrate how the EKP tools can interact with models in a non-intrusive fashion. In particular, this example will be useful if the model is written in another language, requires HPC management, or requies additional data processing stages.

> After instantiation, this example calibration is called from a simple script-language (bash) by calling

Reviewer: ```suggestion
We can generally orchestrate the different stages of the calibration process using a script file. Here we employ a Linux bash script:
```

> After instantiation, this example calibration is called from a simple script-language (bash) by calling

Author: This is what I meant. Removed.

```
> bash calibrate_script
```

> Which calls the following sequence of runs (here in pseudo-code)

Reviewer: More than pseudocode, the loop syntax is in Julia, right? That can be confusing. Why not use bash syntax? 

The fact that we are running the computer model with `Julia --project` is also quite specific. We can leave that if we comment that this need not be a Julia call (for instance, it could be python run_computer_model.py)

> Which calls the following sequence of runs (here in pseudo-code)

Author: I'll give this piece another go, originally I wanted something easy to read but which showed that which files we called where, in the example

# the EKI loop
for i in 1:N_iterations 

    # run the model at each ensemble member
    for j in 1:N_ensemble
        julia --project run_computer_model.jl i j
    end

    # update the ensemble with EKI
    julia --project update_EKP.jl i
    
end
```
The interaction between the calibration tools and the forward map are only through simple readable files stored in a nested file structure: for iteration `i`, ensemble member `j`
- Parameters values are stored (with their priors) in `output/iteration_i/member_j/parameters.toml`
- Computer model outputs are stored in `output/iteration_i/member_j/model_output.jld2`

# Inputs and Outputs

The prior distributions are provided in `priors.toml` in TOML format.
```toml
[amplitude]
prior = "Parameterized(Normal(0.5815754049028404, 0.47238072707743883))"
constraint = "bounded_below(0.0)"
description = """
The amplitude of the sine curve.
This yields a physical prior that is log-normal with approximate (mean,sd) = (2,1)
"""

[vert_shift]
prior = "Parameterized(Normal(0,5))"
constraint = "no_constraint()"
description = """
The vertical shift of the sine curve.
This yields a physical prior that is Normal with (mean,sd) = (0,5)
"""
```

> More information on the priors and constraints are given [here](@ref parameter-distributions). More examples of defining priors in a TOML format are found `test/TOMLInterface/toml/`.

Reviewer: ```suggestion
More information on the priors and constraints are given [here](@ref parameter-distributions). More examples defining priors in TOML format may be found in `test/TOMLInterface/toml/`.
```

After running the example, (it takes *several minutes* to run), results are stored in `output/eki.jld2`. To view the results, one can interact with the stored objects by loading `julia --project` and proceeding as follows:
```julia
julia> using EnsembleKalmanProcesses, JLD2

julia> @load "output/eki.jld2"
3-element Vector{Symbol}:
 :eki
 :param_dict
 :prior
```
Then, for example, the final 5-member parameter ensemble is retrieved with:
```julia
julia> get_ϕ_final(prior,eki)
2×5 Matrix{Float64}:
 1.31749  1.28875  1.27677  1.33311  1.3305
 6.27765  6.63572  6.65148  6.66096  6.56518
```

!!! note "Why is it so slow?"

>     The example is slow because the forward map is written in Julia, and so 99.99% of computation for each call to `julia --project` is precompilation. Ensembles in Julia can be accelerated by using methods discussed [here](@ref parallel-hpc), or by compiling system images with `PackageCompiler.jl`, for example. This example is for instructional purposes only, and so is not optimized.

Reviewer: Is Plots what is taking so long to compile?

Author: No, none of the files call plots with the using command,  I've now removed plots the example project too but this makes no difference. It takes around `14s` to precompile .

Author: It might be `Distributions`?
